### ISSUE TITLE
Accuracy suffers as number of MPI processes increases

### ISSUE THREAD

**Issue by timmoon10** [LLNL/lbann#12]
When I run lbann_dnn_mnist, I find that the accuracy is dependent on the number of MPI ranks. The mini-batch size is 100 and the other parameters are default. With one process, we get 90.52% test accuracy after one epoch. With 16 processes, the test accuracy drops to 80.75%. Looking at the past commits, the error first appears in commit 6be4b032249d47a116adff05c5e48984f387c52f or commit a6adcf3d0df021696ced665799f76abaf542d543.

**Comment by ndryden** [comment]
I continue to see this issue, but the drop in accuracy is not as dramatic.

**Comment by ndryden** [comment]
I double-checked this with the latest version of the code (with `SEQ_INIT=ON`), and it is still present. `./run_lbann_dnn_mnist -u -b 100` gives 93.14% test accuracy on one process and 92.86% on 16 processes, both after five epochs. Note, both have the same accuracy, 89.84%, after one epoch.

**Comment by bvanessen** [comment]
Interesting. Since you are not using the multi-model this bug is different from the one that I previously fixed. It would be good to check if this bug existed before we switched to the new metrics and objective functions. Those do calculate accuracy slightly differently than the previous implementation.



Brian Van Essen
vanessen1@llnl.gov
(w) 925-422-9300
(c) ‭925-290-5470‬
________________________________

**Comment by timmoon10** [comment]
With `SEQ_INIT=OFF` on the single-model fully-connected MNIST model (using the same parameters):

Processes | Epoch 1 Test Accuracy | Epoch 5 Test Accuracy | Epoch 20 Test Accuracy
----|----|----|---
1 | 90.45% | 93.15% | 95.17%
4 | 89.71% | 93.13% | 95.19%
16 | 90.15% | 93.62% | 95.43%

**Comment by timmoon10** [comment]
I don't see this issue when I run on catalyst. With `SEQ_INIT=ON`:

Processes | Epoch 1 Test Accuracy | Epoch 5 Test Accuracy | Epoch 20 Test Accuracy
----|----|----|----
1 | 89.84% | 93.11% | 95.3%
4 | 89.84% | 93.14% | 95.1%
16 | 89.84% | 93.47% | 95.42%

Experiment parameters:
- Run on catalyst
- Mini-batch size: 100
- Adagrad
- Learning rate: 0.01
- Learning rate decay rate: 0.5
- Sigmoid activations
- No dropout
- No L2 regularization

I get nearly identical results when I use one node with 16 processes or four nodes with four processes each.

**Comment by bvanessen** [comment]
Can you please recompile with doubles and run both tests again.

Thanks

Brian C. Van Essen
vanessen1@llnl.gov
(w) 925-422-9300
(c) 925-290-5470

**Comment by JaeseungYeom** [comment]
On surface using 32 ranks,
89.42%@1 epoch, 92.99%@5 epochs, 95.35%@20 epochs with SEQ_INIT=off
89.84%@1 epoch, 93.01%@5 epochs, 95.31%@20 epochs with SEQ_INIT=on

**Comment by timmoon10** [comment]
Repeating my experiment with `SEQ=ON` and double-precision datatype:

Processes | Epoch 1 Test Accuracy | Epoch 5 Test Accuracy | Epoch 20 Test Accuracy
----|----|----|----
1 | 90.41% | 93.09% | 95.27%
4 | 90.41% | 93.14% | 95.37%
16 | 90.41% | 93.33% | 95.39%

**Comment by ndryden** [comment]
Using `run_lbann_dnn_mnist.sh -u -b 100 -e 20` on Catalyst:

Processes | DataType | Epoch 1 Test Accuracy | Epoch 5 Test Accuracy | Epoch 20 Test Accuracy
----|----|----|----|----
1 | float | 89.84% | 93.14% | 95.34%
16 | float | 89.84% | 92.86% | 95.36%
1 | double | 90.41% | 93.04% | 95.25%
16 | double | 90.41% | 92.99% | 95.32%

It's curious that we Tim and I both get the same epoch 1 test accuracy, but epochs 5 and 20 differ.

**Comment by ndryden** [comment]
After some playing around, the issue seems to stem from using shuffling in the distributed_minibatch_parallel_io layers.

**Comment by bvanessen** [comment]
Okay I will look at that.



Brian Van Essen
vanessen1@llnl.gov
(w) 925-422-9300
(c) ‭925-290-5470‬
________________________________

**Comment by bvanessen** [comment]
The source of the variability was that each rank used the same random number generator for initializing matrices as well as shuffling training data sequences.  I have added a separate RNG that is only used for shuffling training/testing/validation data sequences.  Initializing this with a common seed will provide consistent training results independent of the number of MPI processes used.  Committed at https://github.com/LLNL/lbann/commit/509b23cf561377e757a88390e4f194195524bea2
